### PR TITLE
Firmware for 3x3macropad

### PR DESCRIPTION
**Pull request by rarepotato8de**
## Description

Adds support for 3x3macropad.
This is a default firmware with bongocat gif for a DIY open-source 3x3macropad.
OLED and rotary encoder can be customized in the keymap. The macropad is linked in the readme.md.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/rarepotato8de/3x3macropad/config.h
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/config.h:1]: This file should be removed.

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{
+    "manufacturer": "RarePotato8DE",
+    "keyboard_name": "3x3macropad",
+    "maintainer": "RarePotato8DE",
+    "bootloader": "caterina",
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:5]:
```suggestion
    "development_board": "promicro",
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    "build": {
+        "lto": true
+    },
+    "debounce": 5,
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:9]:
```suggestion
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    "features": {
+        "bootmagic": true,
+        "combo": false,
+        "encoder": true,
+        "extrakey": true,
+        "key_override": false,
+        "nkro": true,
+        "oled": true,
+        "rgblight": true,
+        "tap_dance": false
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json]:
```suggestion
        "encoder": true,
        "extrakey": true,
        "nkro": true,
        "oled": true,
        "rgblight": true
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        ]
+    },
+    "features": {
+        "bootmagic": true,
+        "combo": false,
+        "encoder": true,
+        "extrakey": true,
+        "key_override": false,
+        "nkro": true,
+        "oled": true,
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:22]: If OLEDs are enabled at the keyboard level, there needs to exist some code for them as well at the keyboard level in a `3x3macropad.c` file.

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        "rgblight": true,
+        "tap_dance": false
+    },
+    "matrix_pins": {
+        "cols": ["D7", "B3", "B2"],
+        "rows": ["F7", "B5", "B6"]
+    },
+    "processor": "atmega32u4",
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:29]:
```suggestion
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        "animations": {
+            "alternating": true,
+            "breathing": true,
+            "rainbow_mood": true,
+            "rainbow_swirl": true,
+            "snake": true,
+            "static_gradient": true
+        },
+        "led_count": 6,
+        "pin": "E6"
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:40]:
```suggestion
        "led_count": 6
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        "device_version": "0.0.1",
+        "pid": "0x0001",
+        "vid": "0x5353"
+    },
+    "ws2812": {
+        "pin": "E6"
+    },
+    "layouts": {
+        "LAYOUT": {
+            "c_macro": true,
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:53]:
```suggestion
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            "layout": [
+                {"label": "k00", "matrix": [0, 0], "x": 0, "y": 0, "w": 1},
+                {"label": "k01", "matrix": [0, 1], "x": 1, "y": 0, "w": 1},
+                {"label": "k02", "matrix": [0, 2], "x": 2, "y": 0, "w": 1},
+                {"label": "k10", "matrix": [1, 0], "x": 0, "y": 1, "w": 1},
+                {"label": "k11", "matrix": [1, 1], "x": 1, "y": 1, "w": 1},
+                {"label": "k12", "matrix": [1, 2], "x": 2, "y": 1, "w": 1},
+                {"label": "k20", "matrix": [2, 0], "x": 0, "y": 2, "w": 1},
+                {"label": "k21", "matrix": [2, 1], "x": 1, "y": 2, "w": 1},
+                {"label": "k22", "matrix": [2, 2], "x": 2, "y": 2, "w": 1}
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:63]:
```suggestion
                {"label": "k00", "matrix": [0, 0], "x": 0, "y": 0},
                {"label": "k01", "matrix": [0, 1], "x": 1, "y": 0},
                {"label": "k02", "matrix": [0, 2], "x": 2, "y": 0},
                {"label": "k10", "matrix": [1, 0], "x": 0, "y": 1},
                {"label": "k11", "matrix": [1, 1], "x": 1, "y": 1},
                {"label": "k12", "matrix": [1, 2], "x": 2, "y": 1},
                {"label": "k20", "matrix": [2, 0], "x": 0, "y": 2},
                {"label": "k21", "matrix": [2, 1], "x": 1, "y": 2},
                {"label": "k22", "matrix": [2, 2], "x": 2, "y": 2}
```

File: keyboards/rarepotato8de/3x3macropad/config.h
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/config.h:1]: This file should be removed.

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    "encoder": {
+        "rotary": [
+            {"pin_a": "F4", "pin_b": "B1"}
+        ]
+    },
+    "features": {
+        "bootmagic": true,
+        "encoder": true,
+        "extrakey": true,
+        "key_override": false,
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/info.json:20]:
```suggestion
```

File: keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (index == 0) {
+        if (clockwise) {
+            tap_code(KC_VOLD);
+        } else {
+            tap_code(KC_VOLU);
+        }
+    }
+    return false;
+}
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c:36]:
```suggestion
```

File: keyboards/rarepotato8de/3x3macropad/readme.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
+# 3x3macropad
+
+![3x3macropad](https://raw.githubusercontent.com/rarepotato8de/3x3macropad/main/resources/title.png)
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/readme.md:3]: This image needs to be rehosted on Imgur.

File: keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+#ifdef OLED_ENABLE
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c:38]: As above; this should probably be moved to keyboard level.

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/rarepotato8de/3x3macropad/3x3macropad.c [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
+0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+    };
+	oled_write_raw_P(data, sizeof(data));
+}
+
+/* This code switches the shown image on the oled on every 350th call of this function */
+bool showBongo1 = true;
+int calls = 0;
+bool oled_task_user(void) {
```
Comment (zvecr) [inline keyboards/rarepotato8de/3x3macropad/3x3macropad.c:99]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) { return false; }
```

File: keyboards/rarepotato8de/3x3macropad/readme.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# 3x3macropad
+
+![3x3macropad](https://i.imgur.com/OTb3CUq.png)
```
Comment (zvecr) [inline keyboards/rarepotato8de/3x3macropad/readme.md:3]: The image here doesnt match up with the layout macro configured and default keymap

File: keyboards/rarepotato8de/3x3macropad/3x3macropad.c [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
```
Comment (waffle87) [inline keyboards/rarepotato8de/3x3macropad/3x3macropad.c:17]:
```suggestion
#include "quantum.h"
```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    "usb": {
+        "device_version": "0.0.1",
+        "pid": "0x0001",
+        "vid": "0x5353"
+    },
+    "ws2812": {
+        "pin": "E6"
+    },
+    "layouts": {
+        "LAYOUT": {
```
Comment (dunk2k) [inline keyboards/rarepotato8de/3x3macropad/info.json:48]:
```suggestion
    "community_layouts": ["ortho_3x3"],
    "layouts": {
        "LAYOUT_ortho_3x3": {
```

File: keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
```
Comment (dunk2k) [inline keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c:20]:
```suggestion
    [0] = LAYOUT_ortho_3x3(
```

File: keyboards/rarepotato8de/3x3macropad/3x3macropad.c [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+		calls = 0;
+	}
+	if(showBongo1) {
+		render_bongo1();	
+	} else {
+		render_bongo2();
+	}
+    return false;
+}
+#endif
```
Comment (dunk2k) [inline keyboards/rarepotato8de/3x3macropad/3x3macropad.c:114]:
```suggestion
#endif

```

File: keyboards/rarepotato8de/3x3macropad/info.json [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+                {"label": "k10", "matrix": [1, 0], "x": 0, "y": 1},
+                {"label": "k11", "matrix": [1, 1], "x": 1, "y": 1},
+                {"label": "k12", "matrix": [1, 2], "x": 2, "y": 1},
+                {"label": "k20", "matrix": [2, 0], "x": 0, "y": 2},
+                {"label": "k21", "matrix": [2, 1], "x": 1, "y": 2},
+                {"label": "k22", "matrix": [2, 2], "x": 2, "y": 2}
+            ]
+        }
+    }
+}
```
Comment (dunk2k) [inline keyboards/rarepotato8de/3x3macropad/info.json:62]:
```suggestion
}

```

File: keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+        KC_1,   KC_2,   KC_3,
+        KC_4,   KC_5,   KC_6,
+        KC_7,   KC_8,   KC_9
+    )
+};
```
Comment (dunk2k) [inline keyboards/rarepotato8de/3x3macropad/keymaps/default/keymap.c:25]:
```suggestion
};

```

**Review by dunk2k — CHANGES_REQUESTED** [review]
- add Community Layout support
- add EoF, where missing

File: keyboards/rarepotato8de/3x3macropad/3x3macropad.c
Comment (drashna) [inline keyboards/rarepotato8de/3x3macropad/3x3macropad.c:1]: Could you run a format pass here?  Or at least, replace the tabs with spaces?

**Review by drashna — APPROVED** [review]
Aside from the nitpick, looks good.

File: keyboards/rarepotato8de/3x3macropad/3x3macropad.c
Comment (rarepotato8de) [reply keyboards/rarepotato8de/3x3macropad/3x3macropad.c:1]: done. Thanks :)

**Pull request merged**